### PR TITLE
Feat: assignment 6 oops concept

### PR DESCRIPTION
**Pull request by aristovinceselvanathan**

### CODE REVIEW

File: src/Assignment6/Shape_Hierarchy/Shape.cs [hunk 1]
```diff
@@ -0,0 +1,56 @@
+﻿namespace Shape_Hierarchy
+{
+    /// <summary>
+    /// Shape abstract class
+    /// </summary>
+    public abstract class Shape
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Shape"/> class.
+        /// It sets the colour, Input1 and Input2 of the dimensions of the shapes
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Shape.cs]: Remove this line. It is unnecessary. The param descriptions will take care of describing the parameters.

File: src/Assignment6/Shape_Hierarchy/Shape.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        /// </summary>
+        /// <param name="colour">It takes the colour as a string</param>
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Shape.cs]:
```suggestion
        /// <param name="colour">Color of the shape.</param>
```

Update all the param and method descriptions similar to this.

File: src/Assignment6/Shape_Hierarchy/Rectangle.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+﻿namespace Shape_Hierarchy
+{
+    /// <summary>
+    /// Rectangle class that inherits the shape class
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Rectangle.cs]: The class description should not contain info about the base class.

File: src/Assignment6/Shape_Hierarchy/Rectangle.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        /// It calls the base calls and pass the color, length, breath to the base class.
+        /// It prints It is a Rectangle Class
+        /// </summary>
+        /// <param name="colour">It takes the colour of the rectangle as a string</param>
+        /// <param name="length"> It takes the length of the rectangle as a double</param>
+        /// <param name="breath">It takes the breath of the rectangle as a double</param>
+        public Rectangle(string colour, double length, double breath)
+            : base(colour, length, breath)
+        {
+            Console.WriteLine("It is a Rectangle Class");
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Rectangle.cs]: For what this console message is needed?

File: src/Assignment6/Shape_Hierarchy/Shape.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            this.Colour = colour;
+            this.Input1 = input1;
+            this.Input2 = input2;
+        }
+
+        /// <summary>
+        /// Gets or sets the Colour of the shape
+        /// </summary>
+        /// <value>string</value>
+        protected string Colour { get; set; }
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Shape.cs]: Should this be `Public`?

File: src/Assignment6/Shape_Hierarchy/Program.cs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    internal class Program
+    {
+        /// <summary>
+        /// Main Method takes the colour, radius for the circle and colour, length, breath for the rectangle shapes
+        /// It prints the details of each shape
+        /// </summary>
+        /// <param name="args">It takes the string array from the command line interface</param>
+        public static void Main(string[] args)
+        {
+            string colorOfCircle, colorOfRectangle, isDoubleOfRadius, isDoubleOfRectangle, isDoubleOfBreath;
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Program.cs]:
```suggestion
            string colorOfCircle, colorOfRectangle, isDoubleOfRadius, isDoubleOfRectangle, isDoubleOfBreadth;
```

Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Program.cs]: isDoubleOfRadius, isDoubleOfRectangle, isDoubleOfBreath

These variable names are not meaningful. Please update it.

File: src/Assignment6/Shape_Hierarchy/Program.cs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+                InvalidWarning("Color of the Rectangle");
+            }
+        }
+
+        /// <summary>
+        /// Method checks for the string is contains only alphabets
+        /// </summary>
+        /// <param name="s">It takes the name of colour of the shape as input</param>
+        /// <returns>It returns bool about the match the pattern</returns>
+        public static bool IsColorName(string s)
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Program.cs]: IsColorName, InvalidWarning

These function names are not meaningful. Please update it.

File: src/Assignment6/Shape_Hierarchy/Program.cs [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        {
+            Regex r = new Regex("^[a-zA-Z\\s]+$");
```
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Program.cs]: It's good that you've validated this.

File: src/Assignment6/Shape_Hierarchy/Shape_Hierarchy.csproj
Comment (sandeep-sol) [inline src/Assignment6/Shape_Hierarchy/Shape_Hierarchy.csproj:1]: Don't use underscore in-between folder names, class names, and namespaces.

Update it in all the places.

File: src/Assignment6/Employee_Hierarchy/Employee.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    /// </summary>
+    internal abstract class Employee
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Employee"/> class.
+        /// Method sets the name and salary of the object
+        /// </summary>
+        /// <param name="name">It takes the name of the employee as string</param>
+        /// <param name="salary">It takes the salary of the employee as decimal</param>
+        public Employee(string name, decimal salary)
```
Comment (sandeep-sol) [inline src/Assignment6/Employee_Hierarchy/Employee.cs]: Why decimal?

File: src/Assignment6/Employee_Hierarchy/Program.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                isValidDecimal = Console.ReadLine();
+
+                if (decimal.TryParse(isValidDecimal, out salary))
+                {
+                    Console.WriteLine("Did You want to create the 1.Developer 2.Manager 3.Exit : ");
+                    if (int.TryParse(Console.ReadLine(), out optionOfTypeOfRole))
+                    {
+                        switch (optionOfTypeOfRole)
+                        {
+                            case 1:
```
Comment (sandeep-sol) [inline src/Assignment6/Employee_Hierarchy/Program.cs]: Should use enum here.

File: src/Assignment6/Employee_Hierarchy/Program.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        /// <summary>
+        /// Main method it ask for the Name, Salary and type of the role.
+        /// It will print the details about the role.
+        /// </summary>
+        /// <param name="args">It takes the argument from command line</param>
+        public static void Main(string[] args)
+        {
+            int optionOfTypeOfRole;
+            string isValidName, isValidDecimal;
```
Comment (sandeep-sol) [inline src/Assignment6/Employee_Hierarchy/Program.cs]: Shouldn't these be of type `bool`?

File: src/Assignment6/Employee_Hierarchy/Program.cs [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+            if (IsNameValid(isValidName))
+            {
+                Console.WriteLine("Salary of the Employee : ");
+                isValidDecimal = Console.ReadLine();
+
+                if (decimal.TryParse(isValidDecimal, out salary))
+                {
+                    Console.WriteLine("Did You want to create the 1.Developer 2.Manager 3.Exit : ");
+                    if (int.TryParse(Console.ReadLine(), out optionOfTypeOfRole))
```
Comment (sandeep-sol) [inline src/Assignment6/Employee_Hierarchy/Program.cs]: Logically, this should have been the first input.

File: src/Assignment6/Employee_Hierarchy/Program.cs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+
+        /// <summary>
+        /// It creates the role of the employee and the print the details of the employee
+        /// </summary>
+        /// <param name="employee">It takes the employee as the input</param>
+        public static void CreateRoleOfEmployee(Employee employee)
+        {
+            string optionOfChoice;
+
+            Console.WriteLine("Did you want print the details : Y or N");
```
Comment (sandeep-sol) [inline src/Assignment6/Employee_Hierarchy/Program.cs]: Why are these console messages in past tense?

File: src/Assignment6/Banking_System/BankAccount.cs [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        /// </summary>
+        /// <param name="nameOfInput">It takes the name of the input</param>
+        public static void FailedWarning(string nameOfInput)
+        {
+            Console.BackgroundColor = ConsoleColor.Red;
+            Console.ForegroundColor = ConsoleColor.Black;
+            Console.WriteLine($"Invalid {nameOfInput}");
+            Console.ForegroundColor = ConsoleColor.White;
+            Console.BackgroundColor = ConsoleColor.Black;
+        }
```
Comment (sandeep-sol) [inline src/Assignment6/Banking_System/BankAccount.cs]: These functions should not be a part of this class. Please move them.

And also the names are not meaningful. Please update them.

File: src/Assignment6/Banking_System/BankAccount.cs [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+        public virtual void Withdraw(decimal amount)
+        {
+            if (this.Balance - amount < 0)
+            {
+                FailedWarning("Transaction is Failed! Insufficient Funds");
+            }
+            else
+            {
+                this.Balance -= amount;
+            }
```
Comment (sandeep-sol) [inline src/Assignment6/Banking_System/BankAccount.cs]: Is this implementation necessary?

Applicable to all the 3 projects.

File: src/Assignment6/Banking_System/Program.cs [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+                Console.Clear();
+            }
+        }
+
+        /// <summary>
+        /// Method checks for the string is number or not because account number consists of numbers
+        /// </summary>
+        /// <param name="s">It takes the string as input</param>
+        /// <returns>It returns bool</returns>
+        public static bool IsNumberValid(string s)
```
Comment (sandeep-sol) [inline src/Assignment6/Banking_System/Program.cs]: The method name can be more specific.

**Review by sandeep-sol — CHANGES_REQUESTED** [review]

File: src/Assignment6/Shape_Hierarchy/Shape.cs (hunk 1 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Shape.cs]: I have changed it.

File: src/Assignment6/Shape_Hierarchy/Shape.cs (hunk 2 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Shape.cs]: I have resolved it.

File: src/Assignment6/Shape_Hierarchy/Rectangle.cs (hunk 3 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Rectangle.cs]: I have changed the base classes description.

File: src/Assignment6/Shape_Hierarchy/Rectangle.cs (hunk 4 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Rectangle.cs]: I have removed the message.

File: src/Assignment6/Shape_Hierarchy/Shape.cs (hunk 5 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Shape.cs]: I have made to Public.

File: src/Assignment6/Shape_Hierarchy/Program.cs (hunk 6 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Program.cs]: I have changed the variable names.

Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Program.cs]: Variable names are updated.

File: src/Assignment6/Shape_Hierarchy/Program.cs (hunk 7 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Program.cs]: Function name is updated

File: src/Assignment6/Shape_Hierarchy/Shape_Hierarchy.csproj
Comment (aristovinceselvanathan) [reply src/Assignment6/Shape_Hierarchy/Shape_Hierarchy.csproj:1]: I have changed it

File: src/Assignment6/Employee_Hierarchy/Employee.cs (hunk 9 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Employee_Hierarchy/Employee.cs]: I have changed it

File: src/Assignment6/Employee_Hierarchy/Program.cs (hunk 10 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Employee_Hierarchy/Program.cs]: I have used the enum in that switch case

File: src/Assignment6/Employee_Hierarchy/Program.cs (hunk 11 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Employee_Hierarchy/Program.cs]: I have updated the one.

File: src/Assignment6/Employee_Hierarchy/Program.cs (hunk 12 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Employee_Hierarchy/Program.cs]: I have changed the logic

File: src/Assignment6/Employee_Hierarchy/Program.cs (hunk 13 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Employee_Hierarchy/Program.cs]: I have changed the console messages

File: src/Assignment6/Banking_System/BankAccount.cs (hunk 14 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Banking_System/BankAccount.cs]: I have moved the logic

File: src/Assignment6/Banking_System/Program.cs (hunk 16 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/Banking_System/Program.cs]: I have changed the method name

File: src/Assignment1/Assignment1.csproj [hunk 17]
```diff
@@ -14,4 +14,8 @@
     </PackageReference>
   </ItemGroup>
 
+  <ItemGroup>
```
Comment (sandeep-sol) [inline src/Assignment1/Assignment1.csproj]: Why are these changes added here?

Same for Assignment2 and Assignment4 projects.

File: src/Assignment6/BankingSystem/Program.cs [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+                    WarningMessageFromConsole("Option");
+                }
+            }
+        }
+
+        /// <summary>
+        /// This function performs the deposit and withdraw function
+        /// </summary>
+        /// <param name="user">user is reference of the bank account class can have both the child class</param>
+        public static void Services(BankAccount user)
```
Comment (sandeep-sol) [inline src/Assignment6/BankingSystem/Program.cs]:
```suggestion
        public static void Services(BankAccount userAccount)
```

**Review by sandeep-sol — APPROVED** [review]

File: src/Assignment6/BankingSystem/Program.cs (hunk 18 above)
Comment (aristovinceselvanathan) [reply src/Assignment6/BankingSystem/Program.cs]: I have resolved it.

**Pull request merged**